### DISCUSSION TITLE
How can I create a spectrum visualization?

### DISCUSSION THREAD

**Discussion by alinnert** [katspaugh/wavesurfer.js#3773] [Q&A] [answered]
I haven't done anything with audio before but I'm kind of interested in this stuff. What I'd like to make is some sort of visualization while some audio is playing. The starting point for doing something like this should be a spectrum graph, right? As far as I understand this the spectrograph plugin already uses the data I'm looking for but I don't think I can access the raw data. It just outputs the generated graph and that's it.

If I was able to access the raw data I could display it in any shape or form. Is there currently a possibility to do this? Or could this be added to the spectrograph plugin - or a separate one if it makes more sense? Or is this out of scope for this library?

**Comment by entonbiba** [comment]
@alinnert you can customize the colours of the waveform, here is an example of that.
https://wavesurfer.xyz/example/gradient-fill-styles/

For spectrum customizing it's also possible, have a look at the example here:
https://wavesurfer.xyz/example/spectrogram/

For realtime custom animations based on the peaks data you can also get the peaks data via the following and then use your own canvas animations etc:
```
// getPeaks(length, starttime, endtime)
wavesurfer.backend.getPeaks(550,0,550)
```

**Reply by alinnert** [reply]
Okay, the last point is potentially what I'm looking for. But what is `wavesurfer` referring to here? I've tried the following:

```js
import WaveSurfer from 'wavesurfer.js'

WaveSurfer.backend
WaveSurfer.create({...}).backend
```

But none of those do have a `backend` property. Searching for "backend" or "getPeaks" in the docs doesn't get me results either. Also, I couldn't find `getPeaks` when I searched through the code of this repo. Where is it coming from?

**Reply by entonbiba** [reply]
@alinnert you can access the `backend` after its created. Do you have a link to a webpage you are trying to access it or is it in nodejs backend? I'll spin up a sample soon,

**Reply by katspaugh** [reply]
@entonbiba "backend" was available only in wavesurfer v6. In v7, [which was released almost exactly 1 year ago](https://github.com/katspaugh/wavesurfer.js/releases/tag/7.0.0), there's no "backend".

It's possible to get the generated peaks via [`wavesurfer.getDecodedData`](https://wavesurfer.xyz/docs/classes/wavesurfer.default#getDecodedData) if that's what the OP is after.

**Reply by alinnert** [reply, marked as answer]
Alright, thank you a lot! I got a lot further now. So, `wavesurfer.getDecodedData()` gives me the waveform data points, right? They are not exactly what I need. But I was able to find the relevant part of the source that I'm looking for. The spectrogram plugin has a private function `getFrequencies()` which extracts frequency data from the `AudioBuffer` that I get from `getDecodedData()`. I'm interested in that frequency data. Unfortunately, the result doesn't get stored anywhere. Instead, it's used inline to render the spectrogram.

I now realize there's a chance I don't even need wavesurfer for what I'm trying to do. All I need is an `AudioBuffer` (which is a browser API) and calculate frequency data from it (basically what `getFrequencies()` does). I need to look deeper into how that works but it doesn't look impossibly complex - and I guess that's a topic that's very well documented in many places.
